### ISSUE TITLE
download tutorials

### ISSUE THREAD

**Issue by byersiiasa** [iiasa/message_ix#34]
tried to download the tutorials from command line using 
`messageix-dl --local_path c:/github/msgix_tutorials`

Files downloaded correctly to the folder, but threw this error.

```
INFO:root:Retrieving https://github.com/iiasa/message_ix/archive/master.zip
INFO:root:Unzipping C:\Temp\dvile4p7\master.zip to C:\Temp\dvile4p7
INFO:root:Copying C:\Temp\dvile4p7/message_ix-master/tutorial to c:/github/msgix_tutorials/tutorial
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\README.md (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria.png (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_load_scenario.ipynb (overwrite is
 ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_load_scenario_R.ipynb (overwrite
is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_multiple_policies-answers.ipynb (
overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_multiple_policies.ipynb (overwrit
e is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_reticulate.ipynb (overwrite is ON
)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_single_policy.ipynb (overwrite is
 ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\tools.py (overwrite is ON)
Traceback (most recent call last):
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 101, in do_dl
    shutil.rmtree(tmp)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Temp
\\dvile4p7\\master.zip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\Scripts\messageix-dl-script.py", line 10, in <module>
    sys.exit(dl())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 119, in dl
    local_path=args.local_path)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 103, in do_dl
    shutil.rmtree(tmp)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Temp
\\dvile4p7\\master.zip'
PS H:\> messageix-dl --local_path c:/github/msgix_tutorials
INFO:root:Retrieving https://github.com/iiasa/message_ix/archive/master.zip
INFO:root:Unzipping C:\Temp\2net5aru\master.zip to C:\Temp\2net5aru
INFO:root:Copying C:\Temp\2net5aru/message_ix-master/tutorial to c:/github/msgix_tutorials/tutorial
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\README.md (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria.png (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_load_scenario.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_load_scenario_R.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_multiple_policies-answers.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_multiple_policies.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_reticulate.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\austria_single_policy.ipynb (overwrite is ON)
INFO:root:Writing to c:/github/msgix_tutorials/tutorial\Austrian_energy_system\tools.py (overwrite is ON)
Traceback (most recent call last):
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 101, in do_dl
    shutil.rmtree(tmp)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Temp\\2net5aru\\master.zi
p'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\Scripts\messageix-dl-script.py", line 10, in <module>
    sys.exit(dl())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 119, in dl
    local_path=args.local_path)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\site-packages\message_ix\cli.py", line 103, in do_dl
    shutil.rmtree(tmp)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\byers\AppData\Local\Continuum\anaconda3\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Temp\\2net5aru\\master.zi
p'
```

**Comment by gidden** [comment]
I think this can be closed now, @byersiiasa can you confirm please?

**Comment by Jihoon** [comment]
I also saw this. I am just ignoring it for now.

**Comment by khaeru** [comment]
This is probably because archive.close() is not called before shutil.rmtree(tmp) in message_ix.cli.do_dl(). From the [Python docs](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.close):

> You must call close() before exiting your program or essential records will not be written.

…makes it seem like the file handle for `master.zip` will not be released until that call is made. [ZipFile.open() can also be used as a context manager](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.open), which will do this automatically.

**Comment by gidden** [comment]
Yes this seems a  correct diagnosis for me. Sadly it's not consistent
across oses. To close this, we should do the fix and add a test.

**Comment by khaeru** [comment]
@byersiiasa I believe I fixed this via #256; the `.close()` call is now included: https://github.com/iiasa/message_ix/blob/f1a43f017b42c592eb76a88043cca6a2f816c77e/message_ix/cli.py#L101-L102

Can you please test when you've a spare moment, and confirm by closing this issue?

**Comment by byersiiasa** [comment]
fixed!
